[FLINK-10822] Make MetricFetcher update interval configurable

tillrohrmann opened the pull request:
## What is the purpose of the change

Introduce WebOptions.METRIC_FETCHER_UPDATE_INTERVAL configuration option to configure
the update interval of the MetricFetcher.

## Verifying this change

- Added `MetricFetcherTest#testLongUpdateInterval` and `MetricFetcherTest#testShortUpdateInterval`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zentol commented:
Can you update the JIRA reference to FLINK-10822? I already created a JIRA for implementing this a while ago.

zentol commented on docs/_includes/generated/web_configuration.html:
```diff
@@ -52,6 +52,11 @@
             <td>Path to the log file (may be in /log for standalone but under log directory when using YARN).</td>
         </tr>
+        <tr>
+            <td><h5>web.metric-fetcher.update-interval</h5></td>
```
I'm not quite sure whether this is the right place, have you thought about making this a metric option instead? It seems questionable that the MQS and fetcher are configured differently even though they are so tightly connected.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java:
```diff
@@ -38,18 +38,23 @@
 
 	private final File webUiDir;
 
+	private final long metricFetcherUpdateInterval;
```
this doesn't _really_ belong into the handler configuration imo, but we don't have a better place for it do we :/

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java:
```diff
@@ -78,10 +83,17 @@ public static RestHandlerConfiguration fromConfiguration(Configuration configura
+	}
+
+	public long getMetricFetcherUpdateInterval() {
+		return 0L;
```
this must return `metricFetcherUpdateInterval`.

tillrohrmann replied:
Removed this method altogether.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RestHandlerConfiguration.java:
```diff
@@ -78,10 +83,17 @@ public static RestHandlerConfiguration fromConfiguration(Configuration configura
 		final String rootDir = "flink-web-ui";
 		final File webUiDir = new File(configuration.getString(WebOptions.TMP_DIR), rootDir);
 
+		final long metricFetcherUpdateInterval = configuration.getLong(WebOptions.METRIC_FETCHER_UPDATE_INTERVAL);
```
missing sanity checks for being > 0.

tillrohrmann replied:
Added sanity check to `MetricFetcherImpl` constructor

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricFetcherTest.java:
```diff
@@ -184,4 +193,52 @@ public String getValue() {
+	private MetricFetcher<RestfulGateway> createMetricFetcher(long updateInterval, RestfulGateway restfulGateway) {
+		return new MetricFetcher<>(
+			() -> CompletableFuture.completedFuture(restfulGateway),
+			path -> new CompletableFuture<>(),
```
Returning an already completed future should be a safer default implementation.

tillrohrmann replied:
I actually think the opposite is true, because you would not expect that if you get a response, then complete this future that all subsequent responses are completed yet.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricFetcherTest.java:
```diff
@@ -184,4 +193,52 @@ public String getValue() {
+		return new TestingRestfulGateway.Builder()
+			.setRequestMetricQueryServicePathsSupplier(() -> {
+				requestMetricQueryServicePathsCounter.incrementAndGet();
+				return new CompletableFuture<>();
```
Returning an already completed future should be a safer default implementation.

zjffdu commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java:
```diff
@@ -193,7 +193,12 @@ public WebRuntimeMonitor(
+			queryServiceRetriever,
+			scheduledExecutor,
+			timeout,
+			WebOptions.METRIC_FETCHER_UPDATE_INTERVAL.defaultValue());
```
Shouldn't it be the configured value instead of default value ?

tillrohrmann replied:
The `WebRuntimeMonitor` is legacy code which is no longer used.

tillrohrmann commented:
Thanks for the review @zentol and @zjffdu. I've addressed all your comments:

- Renamed `MetricFetcher` into `MetricFetcherImpl` to be able to extract the `MetricFetcher` interface
- Pass in `MetricFetcher` instead of `MetricQueryServiceRetriever` into the `WebMonitorEndpoint`
- Remove `RestHandlerConfiguration#metricFetcherUpdateInterval`
- Moved `METRIC_FETCHER_UPDATE_INTERVAL` into `MetricOptions`

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricFetcherTest.java:
```diff
@@ -184,4 +179,52 @@ public String getValue() {
+	}
+
+	@Test
+	public void testShortUpdateInterval() throws InterruptedException {
```
I have doubts regarding the stability of this test since the MetricFetcher internally relies on `System.currentTimeMillis`, which isn't necessarily as granular as this test requires. I would increase the sleep to 50.

tillrohrmann replied:
Alright, I've hardened the test case with
```
final long start = System.currentTimeMillis();
		long difference = 0L;

		while (difference <= updateInterval) {
			Thread.sleep(2L * updateInterval);
			difference = System.currentTimeMillis() - start;
		}
```

tillrohrmann commented:
Thanks for the review @zentol, I've addressed your comments.

tillrohrmann commented:
Thanks for the review @zentol. Merging once Travis gives green light.

zentol commented:
@tillrohrmann What do you think about disabling the fetcher completely if the interval is configured to 0?

tillrohrmann commented:
@zentol I like the idea. Shall we do this as a follow up since I've just merged the PR.

zentol commented:
I've opened https://issues.apache.org/jira/browse/FLINK-11382 for this.
